Unify currency symbol behavior, decode entities in Currency package

tomalec opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->


In most areas where we use currency symbols, we treat them as text; often, we only decode HTML entities with `decodeEntities`, stripping all HTML. However, in `AdvancedFilters > NumberFilter` we allow HTML. 

I think the use case for the component consumer using  HTML there is not worth the inconsistency, confusion, and problems with such symbols ending up in CSV exports, etc.

Technically, it may be a breaking change for `<NumberFilter>`, but I doubt if such behavior was desired in the first place.
<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Checkout this branch
5. `pnpm run build`
5. Run automated tests
7. Customize currency setup
```php
add_action( 'admin_enqueue_scripts', function () {
	wp_add_inline_script( 'wp-hooks', <<<EOD
	wp.hooks.addFilter( 'woocommerce_admin_report_currency', 'dev-blog-example', function() {
		return {
			code: "NZD",
			decimalSeparator: ".",
			precision: 2,
			priceFormat: "%1\$s %2\$s",
			symbol: "&#x62f;.&#x625;",
			symbolPosition: "left",
			thousandSeparator: ",",
		};
	} );
	EOD );
} );
```
8. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fcustomers&filter=advanced`
9. Add filter: `Total spend`
10. Check that the currency is rendered correctly.

![image](https://github.com/user-attachments/assets/1c432d8e-d4e6-48c6-ab63-241d1e1e4a41)

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Disallow HTML symbols in NumberFilter

tomalec committed:
Add changelog entry

tomalec commented:
The failing E2E seems to be flaky and failing for other PRs as well, like https://github.com/woocommerce/woocommerce/actions/runs/10423642990/job/28872121900

joshuatf reviewed (commented):
Just noting here that some plugins use `woocommerce_currency_symbol` with HTML and this will impact those.  For example, a quick search in GH pulls up [some repos that do this](https://github.com/plugandpay/woocommerce-currency-symbol-gel/blob/master/woocommerce-currency-gel.php).

I believe that's now how this filter is meant to be used based on [this comment](https://github.com/woocommerce/woocommerce/blob/99d8b9bf42a0a949cf65228ed377e277cc945d98/plugins/woocommerce/includes/wc-core-functions.php#L840), but just pointing out that there's not much limiting the use of HTML in the currency symbol elsewhere in WooCommerce.

mikkamp commented:
> Just noting here that some plugins use woocommerce_currency_symbol with HTML and this will impact those. For example, a quick search in GH pulls up [some repos that do this](https://github.com/plugandpay/woocommerce-currency-symbol-gel/blob/master/woocommerce-currency-gel.php).

Thanks @joshuatf  for surfacing those cases. Just for the sake of discussion, I'd like to clarify what the better approach would be. The example that was linked to will still use just a symbol in the backend because of the `is_admin` check. It will only use HTML on the frontend where it includes an icon font.

Also the reasoning we decided to not allow HTML is consistency with other parts of the Analytics / Customers screens. For example if I use a snippet something like this:
```php
add_filter(
	'woocommerce_currency_symbol',
	function ( $currency_symbol, $currency ) {
		if ( 'EUR' === $currency ) {
			$currency_symbol = '<img src="https://cdn.britannica.com/59/236059-050-C8059917/Euro-sign.jpg?w=100">';
		}
		return $currency_symbol;
	},
	10,
	2
);
```

We can see that it's not allowing HTML on all the Analytics screens:
![image](https://github.com/user-attachments/assets/17949f98-52ba-4346-a511-e2640d9d8055)

And while it does allow it to show up in the Customers > Advanced Filter (before applying the patch in this PR), the actual customer report data is broken:
![image](https://github.com/user-attachments/assets/0cd402d7-43e0-4bb6-a50b-0fac133eed5f)

It would again work on the orders screen where it's handled differently (more inconsistency). So what would be the better approach, consistency with the rest of the Customer screen / Analytics or allow sanitized HTML in more locations?

joshuatf commented:
>Also the reasoning we decided to not allow HTML is consistency with other parts of the Analytics / Customers screens.

I agree that we should be consistent with the choices.  If I had to weigh in on which seems more appropriate, I think a symbol and/or 3 letter code is more appropriate, but I'm not overly opinionated on whether this allows HTML or not.   If we choose to not allow HTML, I would expect the same in other cases where the currency symbol is used.

There are a number of other cases like this around WooCommerce which makes me wonder if we choose not to allow HTML, if we should also be coercing properties directly within the function call (`get_woocommerce_currency_symbol`) directly instead of everywhere it's consumed so that we have more consistent results.

> Maybe I'm reading it wrong but from the link here: https://cldr.unicode.org/translation/currency-names-and-symbols
I'm just understanding that a symbol should be used, and if a unique one is not available the 3 letter code should be used instead.

Do you read this as icons or images should be allowed?  My take is that it should be a (unicode) symbol or 3 letter code, but I could be reading this wrong.

mikkamp commented:
> If we choose to not allow HTML, I would expect the same in other cases where the currency symbol is used.

I definitely with that premise. Although that goes far beyond the scope of this PR, that's probably better to pull out into an issue to align currency uses within WooCommerce (or just within the admin).

> if we should also be coercing properties directly within the function call (get_woocommerce_currency_symbol) directly instead of everywhere it's consumed so that we have more consistent results.

That would sound like a simpler approach, do it once for everything. Although that seems to go against the [principle of escaping as late as possible](https://developer.wordpress.org/apis/security/escaping/#toc_3). Although we could definitely add some clarity on what format it's expected to be in so all the consumers of that function know what to expect.

tomalec committed:
Merge branch 'trunk' into update/numberfilter-symbol

tomalec commented:
I did a brief search in the codebase.

I'd like to list a few more places that have already escaped the property. Those that will put HTML markup as plain text and produce odd results.
- https://github.com/woocommerce/woocommerce/blob/8fb0d75fac931f6978d9117d0144dc9cd412b4c0/plugins/woocommerce/includes/admin/meta-boxes/views/html-product-data-variations.php#L207:L207
- https://github.com/woocommerce/woocommerce/blob/dev/deprecate-query/plugins/woocommerce/includes/admin/class-wc-admin-setup-wizard.php#L557:L557
- https://github.com/woocommerce/woocommerce/blob/8fb0d75fac931f6978d9117d0144dc9cd412b4c0/plugins/woocommerce/includes/admin/views/html-bulk-edit-product.php#L41:L41 - put currency symbols in `<input>`'s `placeholder

Also, [this one seems to get broken](https://github.com/woocommerce/woocommerce/blob/dev/deprecate-query/plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js#L468:L468) if we use HTML. It uses `symbol.length` for input fields. The length of HTML markup will not reflect the length of characters in question.

To me, it seems that if we support HTML somewhere, it's not desired in general, and it's better to ensure it does not include HTML early on. 
I propose to escape it straight in https://github.com/woocommerce/woocommerce/blob/8fb0d75fac931f6978d9117d0144dc9cd412b4c0/plugins/woocommerce/includes/wc-core-functions.php#L863:L864. Then if someone puts HTML there, they should spot early it does not work.

Then, escaping it late as well. It will be used as a matter of good practice and security while escaping it early would be just to enforce consistency.

However, this rises a question, should we also escape it in [`wc_currency_settings` ](https://github.com/woocommerce/woocommerce/blob/8fb0d75fac931f6978d9117d0144dc9cd412b4c0/plugins/woocommerce/src/Internal/Admin/Settings.php#L96:L96)? as this is another place, when 3PDs could provide HTML symbol.



-----


Unifying it across the entire WooCommerce is a great idea, but it's beyond the scope of this PR. I'd make a follow-up, and here focus on `@woocommerce/componenets` + `@woocommerce/currency` (as that's what's [fed to props](https://github.com/woocommerce/woocommerce/blob/8fb0d75fac931f6978d9117d0144dc9cd412b4c0/packages/js/components/src/advanced-filters/README.md?plain=1#L73:L73)), and Analytics.

Then, push the change sooner to avoid using the bad practice of inserting markup.

tomalec committed:
Add an explicit note to `CurrencyProps.currency` JSDoc

that HTML markup is not supported.

See the discussion at https://github.com/woocommerce/woocommerce/pull/50726#pullrequestreview-2245838967

eason9487 commented on packages/js/components/src/advanced-filters/number-filter.js:
```diff
@@ -72,16 +73,11 @@ class NumberFilter extends Component {
 		currencySymbol,
 		symbolPosition,
 	} ) {
+		currencySymbol = decodeEntities( currencySymbol );
```
The decoding of HTML entities works well here.

However, if using the example in this PR to set the `symbol` with HTML entities, there are many other places where it won't be rendered properly.

- Such as the Customers table on the same page as in the testing instructions:
   ![1](https://github.com/user-attachments/assets/bbadebb7-b08b-4125-b2d3-5e045b72c64f)
- Most of the analytics pages, like:
   ![image](https://github.com/user-attachments/assets/26f2d459-46d9-48c0-b422-39c0da10130d)

So what I'm trying to figure out is whether the actual use case of allowing HTML entities here would have the same problem? Or is there actually no need to allow HTML entities here?

eason9487 commented on packages/js/components/changelog/update-numberfilter-symbol:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+Unify currency symbol behavior in NumberFilter, disalow HTML.
```
Typo: disalow

eason9487 reviewed (commented):
Left a question about whether HTML entities should be considered as possible use cases.

tomalec committed:
fix typo in changelog

tomalec committed:
Move currency symbol decoding to currency package

Address https://github.com/woocommerce/woocommerce/pull/50726#discussion_r1762988095

tomalec committed:
Add changelog entry for currency package

tomalec commented:
Thank you, @eason9487, for the review!

> However, if using the example in this PR to set the symbol with HTML entities, there are many other places where it won't be rendered properly.

I changed the implementation so it decodes on the currency package level.
It solves all the other uses, as it delivers an already decoded entry to all the components.
It makes the behavior of `setCurrency` consistent with `getDataForCountry`.

eason9487 reviewed (approved):
Thanks for the work! LGTM.
